caas/kubernetes/provider/application/application.go

```diff
@@ -392,6 +388,158 @@ func (a *app) Exists() (caas.DeploymentState, error) {
+// It's only configured once when the application was deployed in the first time.
+func (a *app) configureDefaultService(annotation annotations.Annotation) (err error) {
+	svc := resources.NewService(a.name, a.namespace, &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      a.labels(),
+			Annotations: annotation,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: a.labels(),
+			Type:     corev1.ServiceTypeClusterIP,
+		},
+	})
+	if err = svc.Get(context.Background(), a.client); errors.IsNotFound(err) {
+		return svc.Apply(context.Background(), a.client)
+	}
+	return errors.Trace(err)
+}
+
+// UpdateService updates the default service with specific service type and port mappings.
+func (a *app) UpdateService(param caas.ServiceParam) error {
+	// This method will be used for juju [un]expose.
+	// TODO(embedded): it might be changed later when we have proper modelling for the juju expose for the embedded charms.
+	svc, err := a.getService()
+	if err != nil {
+		return errors.Annotatef(err, "getting existing service %q", a.name)
+	}
+	svc.Service.Spec.Type = corev1.ServiceType(param.Type)
+	svc.Service.Spec.Ports = make([]corev1.ServicePort, len(param.Ports))
+	for i, p := range param.Ports {
+		svc.Service.Spec.Ports[i] = convertServicePort(p)
+	}
+
+	applier := a.newApplier()
+	applier.Apply(svc)
+	if err := a.updateContainerPorts(applier, svc.Service.Spec.Ports); err != nil {
+		return errors.Trace(err)
+	}
+	return applier.Run(context.Background(), a.client, false)
+}
+
+func convertServicePort(p caas.ServicePort) corev1.ServicePort {
+	return corev1.ServicePort{
+		Name:       p.Name,
+		Port:       int32(p.Port),
+		TargetPort: intstr.FromInt(p.TargetPort),
+		Protocol:   corev1.Protocol(p.Protocol),
+	}
+}
+
+func (a *app) getService() (*resources.Service, error) {
+	svc := resources.NewService(a.name, a.namespace, nil)
+	if err := svc.Get(context.Background(), a.client); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return svc, nil
+}
+
+// UpdatePorts updates port mappings on the specified service.
+func (a *app) UpdatePorts(ports []caas.ServicePort) (err error) {
+	svc, err := a.getService()
+	if err != nil {
+		return errors.Annotatef(err, "getting existing service %q", a.name)
+	}
+	svc.Service.Spec.Ports = make([]corev1.ServicePort, len(ports))
+	for i, port := range ports {
+		svc.Service.Spec.Ports[i] = convertServicePort(port)
+	}
+	applier := a.newApplier()
+	applier.Apply(svc)
+	if err = a.updateContainerPorts(applier, svc.Service.Spec.Ports); err != nil {
+		return errors.Trace(err)
+	}
+	err = applier.Run(context.Background(), a.client, false)
+	return nil
+}
+
+func convertContainerPort(p corev1.ServicePort) corev1.ContainerPort {
+	return corev1.ContainerPort{
+		Name:          p.Name,
+		ContainerPort: p.TargetPort.IntVal,
+		Protocol:      corev1.Protocol(p.Protocol),
+	}
+}
+
+func (a *app) updateContainerPorts(applier resources.Applier, ports []corev1.ServicePort) error {
```

Comment:
We don't need to do this because we currently don't support host networking on pods. Considering this would be a rolling update, let's avoid this.

Reply:
yeah, now, we can decide if we need to update container ports now.